File: src/main/java/fortress/tfol/Type.java
package fortress.tfol;

import fortress.util.Errors;
import fortress.tfol.operations.TermVisitor;

/**
* @publish
* Represents a type, sometimes called a sort.
*/
public class Type {
    
    // Published Interface
    
    /**
    * @publish
    * Returns a Type with the given name.
    */
    public static Type mkTypeConst(String name) {
        Errors.failIf(name.length() < 1, "Cannot create type with empty name");
        Errors.failIf(Names.isIllegal(name), "Illegal type name " + name);
        return new Type(name);
    }
    
    // End of Published Interface
    private final String name;
    
    private Type(String name) {
        this.name = name;
    }
    
    public static Type Bool = mkTypeConst("Bool");
    
    public String getName() {
        return name;
    }
    
    @Override
    public String toString() {
        return name;
    }
    
    @Override
    public boolean equals(Object other) {
        // Template method design
        if(this == other) {
            return true;
        }
        if(null == other) {
            return false;
        }
        if(this.getClass() != other.getClass()) {
            return false;
        }
        return this.name.equals( ((Type)other).getName() );
    }
    
    @Override
    public int hashCode() {
        return name.hashCode();
    }
}

File: src/main/scala/fortress/tfol/Type.scala
package fortress.tfol

import fortress.util.Errors

/** Represents a type (or sort if your prefer). */
sealed abstract class Type {
    def name: String
    def getName: String = name
    def isBuiltin: Boolean
    override def toString: String = name
}

case class TypeConst private (name: String) extends Type {
    Errors.precondition(name.length > 0, "Cannot create type with empty name");
    Errors.precondition(! Names.isIllegal(name), "Illegal type name " + name);
    Errors.precondition(! name.contains(" "), "Cannot make type constant with space " + name)
    Errors.precondition(! Type.nameMimicsBuiltin(name), "Cannot make type constant with name " + name)
    
    override def isBuiltin: Boolean = false
}

case object BoolType extends Type {
    def name: String = "Bool"
    override def isBuiltin: Boolean = true
}

case object IntType extends Type {
    def name: String = "Int"
    override def isBuiltin: Boolean = true
}

case object RealType extends Type {
    def name: String = "Real"
    override def isBuiltin: Boolean = true
}

case class BitVectorType(bitwidth: Int) extends Type {
    def name: String = "BitVec " + bitwidth.toString
    override def isBuiltin: Boolean = true
}

object Type {
    def mkTypeConst(name: String): Type = TypeConst(name)
    
    val Bool: Type = BoolType
    val Int: Type = IntType
    val Real: Type = RealType
    def BitVector(bitwidth: Int) = BitVectorType(bitwidth)
    
    def nameMimicsBuiltin(name: String) = {
        name == "Bool" || name == "Int" || name == "Real" || name.startsWith("BitVec")
    }
}
